### PR TITLE
REPL update

### PR DESCRIPTION
**Pull request by Endercheif**
Things added
- Using [curties](https://github.com/bpython/curtsies) for terminal functions
- `<Ctrl-d>` ends properly
- `<Ctrl-l>` clears everything except current line
- Pasting works properly
- Random keystrokes dont break the repl
- <TAB> creates 4 spaces (because spaces > tabs)

### CODE REVIEW

File: src/samarium/shell.py [hunk 1]
```diff
@@ -9,28 +16,129 @@
...
+class FakeIO:
+    def __init__(self) -> None:
+        self.output: list[str] = []
+
+    def write(self, __s: str) -> int:
+        self.output.append(__s)
+        return len(__s)
+
+
+class SamariumRepl:
```
Comment (trag1c) [inline src/samarium/shell.py:28]: I believe this should be named `SamariumREPL` to follow PEP8.

File: src/samarium/shell.py [hunk 2]
```diff
@@ -9,28 +16,129 @@
...
-            print()
-        except EOFError:
-            break
+class FakeIO:
+    def __init__(self) -> None:
+        self.output: list[str] = []
+
+    def write(self, __s: str) -> int:
+        self.output.append(__s)
+        return len(__s)
```
Comment (trag1c) [inline src/samarium/shell.py:25]: Why `__s` and not just `s`?

File: src/samarium/shell.py [hunk 3]
```diff
@@ -9,28 +16,129 @@
...
+        if not statement:
+            return False
+        statement += " ;"
+        error, _ = match_brackets(tokenize(statement))
+
+        if not error:
+            self.execute(statement)
+
+            if self.io.output:
+                self.messages.extend("".join(self.io.output).split("\n"))
```
Comment (trag1c) [inline src/samarium/shell.py:119]:
```suggestion
                self.messages.extend("".join(self.io.output).splitlines())
```

**Review by trag1c — CHANGES_REQUESTED** [review]
Some methods are missing the `-> None` typehint

File: src/samarium/shell.py (hunk 2 above)
Comment (Endercheif) [reply src/samarium/shell.py:25]: Well originally this was subclassing StringIO and when I used intelisense to autocomplete the method it chose that variable name.

File: src/samarium/shell.py (hunk 3 above)
Comment (Endercheif) [reply src/samarium/shell.py:119]: this is intentional. it is not splitting lines if a file but splitting the newlines that written to stdout/stderr.

Comment (trag1c) [reply src/samarium/shell.py:119]: and... ?

**Comment by Endercheif** [comment]
@trag1c fixed

File: src/samarium/shell.py (hunk 3 above)
Comment (Endercheif) [reply src/samarium/shell.py:119]: iirc splitlines can split on things other than `\n`

**Review by trag1c — APPROVED** [review]

**Pull request merged**